### DISCUSSION TITLE
Serialize a json string into a Pydantic model in a multipart Form

### DISCUSSION THREAD

**Discussion by Ran4** [fastapi/fastapi#8976] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

I'm trying to serialize a json string sent in as a multipart form.

This is a minimally reproducible server:

```Python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Spec(BaseModel):
    some_field: str

@app.post("/foo")
async def foo(
        spec: Spec = Form(..., ),
        file: List[UploadFile] = File(...),
        ):
    return "success!"
```

The server will, when called with this request, as sent from python's requests library:

```Python
requests.post(url, data={"spec": spec.json()}, files={"file": (filename, file_handle)})
```

where `spec` is an instance of `Spec` above, return 422 with body

```Python
{"detail":[{"loc":["body","spec"],"msg":"value is not a valid dict","type":"type_error.dict"}]}
```

### Environment

* OS: Ubuntu 20.04
* FastAPI Version 0.61.1
* Python version: 3.8.5

**Comment by Kludex** [comment]
This can help you: https://github.com/tiangolo/fastapi/issues/2387#issuecomment-731662551

**Comment by falkben** [comment]
I feel like I can understand what's going on with the following a bit better, but I believe they are essentially doing the same thing:

```python
def as_form(cls):
    """ decorator for you pydantic model to turn them into form models """
    cls.__signature__ = cls.__signature__.replace(
        parameters=[
            arg.replace(default=Form(...) if arg.default is arg.empty else Form(arg.default))
            for arg in cls.__signature__.parameters.values()
        ]
    )
    return cls
```

then you use it like so:

```python
class UserModel(BaseModel):
    username: str

@service.as_form
class UserModelForm(UserModel):
    """ form model """
```

any path operators that use `UserModelForm` accept form data. any path operators that use `UserModel` accept json

like so (for form data):

```python
@router.post("/post_form")
def post_form(self, user: UserModel = Depends(UserModelForm)):
    return user
```

**Comment by Maksych** [comment, marked as answer]
@Ran4 
I think this solution is what you need.

```python
import fastapi
import pydantic

app = fastapi.FastAPI()


class User(pydantic.BaseModel):
    username: str


@app.post("/create-user-json")
def create_user_json(
    user: pydantic.Json[User] = fastapi.Form(...),
    file: fastapi.UploadFile = fastapi.File(...),
) -> None:
    ...

```

**Comment by dotX12** [comment]
@Ran4 

https://github.com/dotX12/pyfa-converter


```python3
from datetime import datetime
from typing import Optional

from fastapi import FastAPI, UploadFile, File, Form
from pydantic import BaseModel, Field

from pyfa_converter import FormDepends, PyFaDepends

app = FastAPI()


class PostContractBodySchema(BaseModel):
    title: str = Field(..., description="Description title")
    date: Optional[datetime] = Field(
        None, description="Example: 2021-12-14T09:56:31.056Z"
    )


@app.post("/form-data-body")
async def example_foo_body_handler(
    data: PostContractBodySchema = FormDepends(PostContractBodySchema),
    # data1: PostContractBodySchema = PyFaDepends( # OR
    #         model= PostContractBodySchema, _type=Form
    #     ),
    document: UploadFile = File(...),
):
    return {"title": data.title, "date": data.date, "file_name": document.filename}

```
